Added tests and tracking logic for removal before additions finished

grofit opened the pull request:
This exhibits in situations where one tracker kicks off some logic in a system which removes the entity, but other trackers are yet to run their add code but have fired their remove code already, so we basically do the same fix where we check for this then exit if needed as we no longer need to setup.

Bezarius commented:
Use the Ienumerable is acceptable if you want to just run through the array and do some action.  Something like this: 

`foreach(var val in values) { ... }`

in fact it will be deployed as follows:

```
IEnumerable<T> values;
using (var valuesEnum = values.GetEnumerator())
{
    while (valuesEnum .MoveNext())
    {
        // Do something with valuesEnum .Current.
    }
}
```

grofit commented:
Looks good, thanks.

There was historically some issues raised with performance and GC around using foreach style enumeration, I dont know if the MoveNext style approach incurrs the same impact or not, but the normal array index access is one of the quickest ways so as done in the PR would be fine for now. Personally I dont need it to be SUPER performant, the quicker the better but as I said to others I often want to keep functionality paramount and performance secondary.

grofit committed:
Added tests and tracking logic for removal before additions finished

This exhibits in situations where one tracker kicks off some logic in a system which removes the entity, but other trackers are yet to run their add code but have fired their remove code already, so we basically do the same fix where we check for this then exit if needed as we no longer need to setup.
